encode/av1encode.c

```diff
@@ -0,0 +1,2964 @@
+va_render_packed_data(bitstream* bs)
+{
+    CHECK_BS_NULL(bs);
+    VAEncPackedHeaderParameterBuffer packedheader_param_buffer;
+    VABufferID packed_para_bufid = VA_INVALID_ID;
+    VABufferID packed_data_bufid = VA_INVALID_ID;
+    VABufferID render_id[2] = {VA_INVALID_ID};
+    unsigned int length_in_bits = bs->bit_offset;
+
+    unsigned char *packedpic_buffer = bs->buffer;
+    VAStatus va_status;
+
+    packedheader_param_buffer.type = VAEncPackedHeaderPicture; 
+    packedheader_param_buffer.bit_length = length_in_bits;
+    packedheader_param_buffer.has_emulation_bytes = 0;
+
+    va_status = vaCreateBuffer(va_dpy,
+                               context_id,
+                               VAEncPackedHeaderParameterBufferType,
+                               sizeof(packedheader_param_buffer), 1, &packedheader_param_buffer,
+                               &packed_para_bufid);
+    CHECK_VASTATUS(va_status, "vaCreateBuffer");
+
+    va_status = vaCreateBuffer(va_dpy,
+                               context_id,
+                               VAEncPackedHeaderDataBufferType,
+                               (length_in_bits + 7) / 8, 1, packedpic_buffer,
+                               &packed_data_bufid);
+    CHECK_VASTATUS(va_status, "vaCreateBuffer");
+
+    render_id[0] = packed_para_bufid;
+    render_id[1] = packed_data_bufid;
+    va_status = vaRenderPicture(va_dpy, context_id, render_id, 2);
+    CHECK_VASTATUS(va_status, "vaRenderPicture");
+
+    //free(packedpic_buffer); free outside this function call
+
+    if (packed_para_bufid != VA_INVALID_ID) {
+        vaDestroyBuffer(va_dpy, packed_para_bufid);
+        packed_para_bufid = VA_INVALID_ID;
+    }
+    if (packed_data_bufid != VA_INVALID_ID) {
+        vaDestroyBuffer(va_dpy, packed_data_bufid);
+        packed_data_bufid = VA_INVALID_ID;
+    }
+
+}
+
+static void
+render_ivf_frame_header()
+{
+    // write 12 empty byte to driver
+    // IVF frame header is filled after encoding
+    // first 4 byte is u32 for bit stream length
+    // last 8 byte is u64 for display order
+    bitstream bs;
+    bitstream_start(&bs);
+
+    for (size_t i = 0; i < 12; i++)
+    {
+        put_ui(&bs, 0x00, 8);
+    }
+
+    va_render_packed_data(&bs);
+
+    bitstream_free(&bs);
+}
+
+static void
+render_ivf_header()
+{
+    bitstream bs;
+    bitstream_start(&bs);
+    uint32_t ivfSeqHeader[11] = {0x46494B44, 0x00200000, 0x31305641,
+    (uint32_t)(fh.UpscaledWidth + (fh.FrameHeight << 16)),
+    ips.frame_rate_extN,// FrameRateExtN
+    ips.frame_rate_extD,// FrameRateExtD
+    ips.frame_count/*numFramesInFile*/,
+    0x00000000,
+    0,0,0 };
+
+    // 
+    uint8_t* hdr = (uint8_t*) ivfSeqHeader;
+    for (size_t i = 0; i < 44; i++)
+    {
+        put_ui(&bs, hdr[i], 8);
+    }
+
+    va_render_packed_data(&bs);
+
+    bitstream_free(&bs);
+}
+
+static void
+render_TD()
+{
+    // OBU_TEMPORAL_DELIMITER
+    bitstream bs;
+    bitstream_start(&bs);
+    put_ui(&bs, 0x12, 8);
```

Comment:
fix length fix value？

Reply:
This is for OBU_TEMPORAL_DELIMITER. This obu does not have payload, so length is not needed. Value is fixed and calculated directly from av1 spec.